rfc/rfc-73/rfc-73.md

```diff
@@ -0,0 +1,416 @@
+6. **Monitoring:** Integration with tools such as hudi-cli to track the status and health of transactions. Separate
+   transaction metrics. Also, provide comprehensive logs for debugging transaction failures or issues.
+7. **Integration with Current Hudi Features:** The multi-table transaction should be seamlessly integrated with existing
+   Hudi features and should not break existing APIs or workflows. In
+   particular, `TransacationManager`, `LockManager`, `ConflictResolutionStrategy` APIs should work as they do today with
+   single table.
+8. **Configurability:** Allow users to configure the behavior of multi-table transactions, e.g., concurrency control
+   mechanism, timeout durations, etc.
+9. **Recovery**: Rollbacks, savepoint and restore should be supported.
+
+### Non-goals
+
+1. **Cross-Database Transactions:** Transactions spanning multiple databases might be too complex as an initial target.
+2. **Granular Record Locking:** Instead of locking at the granularity of individual records, coarse-grained locking (
+   like at the file level) might be more practical to start with. But, let's avoid table or partition level locking as
+   mentioned in the goals.
+3. **Distributed Transactions:** We are not going to consider transactions in the sense of distributed databases, and
+   thus the need of a protocol such as 2PC or a quorum mechanism to proceed with the commits.
+4. **Complex Conflict Resolution:** Initially, simple conflict resolution strategies (like aborting a transaction on
+   conflict) can be implemented, leaving more sophisticated strategies for future iterations.
+5. **Replication/CDC stream:** We are not going to build a replication or CDC stream out of the database level commit
+   log, though that would be a good usage of the database timeline introduced in this RFC.
+
+### Design
+
+Our primary challenge is to support operations that span multiple tables within a single database, and maintain the ACID
+properties across these tables.
+
+1. Need for a catalog: Do we need a catalog API that tracks databases and its tables?
+2. Need for a transaction coordinator: A centralized coordinator that will manage transaction logs, track ongoing
+   transactions, handle timeouts, and manage transaction rollbacks.
+3. Need for a transaction log: At the table level, the timeline incorporating state of an action with start and modified
+   time serves the purpose of transaction log. At the database level, we need to track all multi-table transactions.
+   Every start, update, or commit of a multi-table transaction gets recorded (Database timeline?).
+4. Locking/Conflict resolution mechanism: Lock the affected files during the multi-table transaction to prevent
+   conflicting writes. Decide on conflict resolution strategies (e.g., last write wins, version vectors).
+5. Need for buffer management: Writes by a transaction are not immediately visible to other transactions. They are
```

Comment:
we cannot write into any temp folders, if that's what you are suggesting. Idk if this is needed, dirty reads are already guarded by table timeline and if a database level timeline can guard dirty reads at multi txn level (i.e prevent commit to a table from being visible until all writes in multi txn complete), we should be good?

Reply:
I am going to remove it. We don't need buffer management. The intention was to have some staging area before flushing data files to storage. That way we can guarantee that data files won't be visible until the all statements of a transaction are completed. But, on another thought, the table timeline together with the database timeline is sufficient to ensure that.

Reply:
yes. this may be confusing to read, can you please reword the bullet title/contents?